packages/design-system/src/components/IdleTimeout/IdleTimeout.tsx

```diff
@@ -77,70 +78,148 @@ const defaultMessageFormatter = (timeTilTimeout: number): React.ReactNode => {
   endSessionButtonText = 'Logout',
+  endSessionRedirectUrl = '/logout',
   formatMessage = defaultMessageFormatter,
-  onClose,
   onSessionContinue,
   onSessionForcedEnd,
   onTimeout,
   showSessionEndButton = false,
   timeToTimeout,
   timeToWarning = 5,
-}: IdleTimeoutProps) => {
-  const handleContinueSession = () => {
-    // reset countdown timer
-    if (onSessionContinue) {
-      onSessionContinue();
+}: IdleTimeoutProps): JSX.Element => {
+  if (timeToWarning > timeToTimeout) {
+    console.error(
+      'Error in TimeoutManager component. `timeToWarning` is greater or equal to `timeToTimeout`'
+    );
+  }
+  // convert minutes to milliseconds
+  const msToTimeout = timeToTimeout * 60000;
+  const msToWarning = (timeToTimeout - timeToWarning) * 60000;
+  const [checkStatusTime, setCheckStatusTime] = useState<number>(null);
+  const [warningIntervalId, setWarningIntervalId] = useState<NodeJS.Timeout>(null);
+  const [showWarning, setShowWarning] = useState<boolean>(false);
+  const [timeInWarning, setTimeInWarning] = useState<number>(timeToWarning);
+
+  // cleanup timeouts & intervals
+  const clearTimeouts = () => {
+    setCheckStatusTime(null);
+    if (warningIntervalId) {
+      clearInterval(warningIntervalId);
     }
   };
 
-  const handleEndSession = () => {
-    if (onSessionForcedEnd) {
-      // TODO: figure out if any params should be passed back to the app
-      onSessionForcedEnd();
-    } else if (onTimeout) {
-      onTimeout();
+  // when the countdown for the session ends, clean up, call callback & close modal
+  const handleTimeout = () => {
+    clearTimeouts();
+    removeEventListeners();
+    onTimeout();
+    setShowWarning(false);
+  };
+
+  // when it's time to warn the user about idleness,
+  // set an interval that updates the modal message
+  const handleWarningTimeout = () => {
+    setShowWarning(true);
+    removeEventListeners();
+    let timeTilTimeout = timeToWarning - 1;
+    const intervalId = setInterval(() => {
+      setTimeInWarning(timeTilTimeout);
+      timeTilTimeout--;
+    }, 60000);
+    setWarningIntervalId(intervalId);
+  };
+
+  const setTimeoutCookies = () => {
+    const timeoutTime = Date.now() + msToTimeout;
+    const warningTime = Date.now() + msToWarning;
+    localStorage.setItem(timeoutCookieName, timeoutTime.toString());
+    localStorage.setItem(timeoutWarningCookieName, warningTime.toString());
+
+    if (checkStatusTime === null) {
+      setCheckStatusTime(30000);
     }
   };
 
-  const handleDialogClose = () => {
-    if (onClose) {
-      onClose();
+  const resetTimeouts = () => {
+    clearTimeouts();
+    setTimeoutCookies();
+  };
+
+  const removeEventListeners = () => {
+    document.removeEventListener('mousemove', resetTimeouts);
+    document.removeEventListener('keypress', resetTimeouts);
+  };
+
+  const addEventListeners = () => {
+    document.addEventListener('mousemove', resetTimeouts);
+    document.addEventListener('keypress', resetTimeouts);
```

Comment:
Did you look into the `passive` option? https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener#parameters

Reply:
I just realized I forgot about this after our conversation 🤦🏻 
The only thing that made me hesitate is that `passive` isn't supported in IE and I don't know if it's through a runtime error in IE if it's present in the code, so I may have to check if there's a way to test if a browser supports it and set that option conditionally. Thoughts?

Reply:
That docs page shows how to test for it. In the absense of any other optimizations like debouncing, it'd probably be better to have this kind of optimization for 99% of our users rather than none of our users. It would be annoying to have to test for support, but at least we could hide away that code in another module.